# FLORENCE

FLORENCE is an R package that provides functions to compute expected values of the site frequency spectrum of somatic variants generated by dividing tissues that are either expanding or homeostatic. Modalities for both, neutrally evolving tissues and tissues with selection of a single clone are provided.

The computations assume deterministic cell numbers in the tissue overall but account for the stochasticity of clonal growth at the level of individual clones, using clone size distributions generated by linear birth-death processes. 

FLORENCE also provides function modalities to simulate VAF distributions generated by bulk whole genome sequencing or by pseudo-bulk data from single-cell WGS. 

## Documentation

Function documentation is available in the accompanying manual (FLORENCE_0.0.0.9000.pdf)[FLORENCE_0.0.0.9000.pdf]. A detailed guide on how to use FLORENCE in conjunction with pyABC is given on github.de:VerenaK90/Clonal_hemopoiesis 

## System requirements

### Hardware requirements

FLORENCE runs on a standard computer.

### Software requirements

#### OS Requirements

This package is supported for macOS and Linux. The package has been tested on the following systems:

macOS: Monterey (12.6.3)
Linux: CentOS Linux 7 (Core)

#### R dependencies

FLORENCE has been tested on R v4.2.0 and v4.2.1 and requires installation of the package deSolve (v1.34).

FLORENCE can be run in conjunction with python3 and pyABC (≥0.12.6) to estimate parameters from whole genome sequencing data.

## Version

0.0.0.9000

## Citation

Körber et al., Quantifying clonal selection and drift from a single bulk tissue sample.

## Installation

library(devtools)
devtools::install.github("VerenaK90/FLORENCE")

## License

## Contact

Verena Körber (v.koerber@dkfz.de)
